qctrlopencontrols/driven_controls/driven_control.py

```diff
@@ -278,25 +278,21 @@ def directions(self):
         amplitudes = np.sqrt(
             self.amplitude_x ** 2 + self.amplitude_y ** 2 + self.detunings ** 2
         )
+
+        amplitudes = np.where(np.isclose(amplitudes, 0.0), 1.0, amplitudes)
```

Comment:
I don't know how relevant this would be here, but if all amplitudes are below 1e-8 (the default `atol` in `np.isclose`) because of the units the user is using, they'll all be flagged as close to 0.

A workaround would be to compare with "normalised amplitudes":
`np.where(np.isclose(amplitudes/np.max(amplitudes), 0.0), 1.0, amplitudes)`
but that would also fail if they're all zero.

If it's a "hard" division by zero you could also check 
`np.where(amplitudes == 0.0, 1.0, amplitudes)`?

Reply:
That's a good point. I was a bit afraid of comparing to zero directly with `== 0.0` (it might not be a problem though, as probably np.sqrt rounds the amplitude down to zero when the numbers get really small, but I'm not completely confident of that), so instead I changed `np.isclose` to use a reduced `atol` that is 1e-20 of the maximum value of the amplitudes.

I'm not convinced that this is the best solution, but I probably won't manage to come up with better ideas today. For what's worth, it seems to work with the extra test I added (which uses ridiculously small amplitudes of the order of 1e-100). Other suggestions are welcome, in any case

Reply:
I like your solution, it's also probably smarter to change the `atol` than normalize the inputs. (I also can't come up with any better alternatives).

Reply:
Note that `durations` are specified to be in seconds, so the user doesn't really have any choice for units (the amplitudes should all be in Hz).

Still, I don't know what the best thing to do is in situations like this. If we use a hard `==0.` check then we'll silently get floating point errors (and possibly `nan`s) when the user gets too close to `0` (which can also happen with the current solution, since the `1e-20` multiplication can send the tolerance down to `0`). If we use a check with a magic number then we'll silently provide an inaccurate value if the user goes too small (but that inaccurate value will still satisfy some nice properties, such as giving a unit vector).

Given all that, I'm fine with this as-is, but I'm also keen to work out some pattern we can use for this sort of thing when we inevitably hit it again.

Reply:
Yeah, I'd be keen to workshop some solution for this kind of situation, as we seem to keep running into that. I imagine other people also have run into this and come up with their own ingenious solutions, and maybe we can learn from them if we figure out where to find them. Unfortunately I ran out of ideas for today